presto-native-execution/presto_cpp/main/PrestoToVeloxQueryConfig.cpp

```diff
@@ -0,0 +1,212 @@
+      },
+
+      {std::string(SystemConfig::kSpillEnabled),
+       velox::core::QueryConfig::kSpillEnabled},
+
+      {std::string(SystemConfig::kJoinSpillEnabled),
+       velox::core::QueryConfig::kJoinSpillEnabled},
+
+      {std::string(SystemConfig::kOrderBySpillEnabled),
+       velox::core::QueryConfig::kOrderBySpillEnabled},
+
+      {std::string(SystemConfig::kAggregationSpillEnabled),
+       velox::core::QueryConfig::kAggregationSpillEnabled},
+
+      {std::string(SystemConfig::kRequestDataSizesMaxWaitSec),
+       velox::core::QueryConfig::kRequestDataSizesMaxWaitSec},
+
+      {std::string(SystemConfig::kDriverMaxSplitPreload),
+       velox::core::QueryConfig::kMaxSplitPreloadPerDriver},
+
+      {std::string(SystemConfig::kMaxLocalExchangePartitionBufferSize),
+       velox::core::QueryConfig::kMaxLocalExchangePartitionBufferSize},
+
+      {std::string(SystemConfig::kUseLegacyArrayAgg),
+       velox::core::QueryConfig::kPrestoArrayAggIgnoreNulls},
+
+      {std::string(SystemConfig::kSinkMaxBufferSize),
+       velox::core::QueryConfig::kMaxOutputBufferSize,
+       [](const auto& value) {
+         return folly::to<std::string>(velox::config::toCapacity(
+             value, velox::config::CapacityUnit::BYTE));
+       }},
+
+      {std::string(SystemConfig::kDriverMaxPagePartitioningBufferSize),
+       velox::core::QueryConfig::kMaxPartitionedOutputBufferSize,
+       [](const auto& value) {
+         return folly::to<std::string>(velox::config::toCapacity(
+             value, velox::config::CapacityUnit::BYTE));
+       }},
+
+      {std::string(SystemConfig::kTaskMaxPartialAggregationMemory),
+       velox::core::QueryConfig::kMaxPartialAggregationMemory,
+       [](const auto& value) {
+         return folly::to<std::string>(velox::config::toCapacity(
+             value, velox::config::CapacityUnit::BYTE));
+       }},
+  };
+
+  for (const auto& configMapping : veloxToPrestoConfigMapping) {
+    const auto& veloxConfigName = configMapping.veloxConfig;
+    const auto& systemConfigName = configMapping.prestoSystemConfig;
+    const auto propertyOpt = systemConfig->optionalProperty(systemConfigName);
+    if (propertyOpt.hasValue()) {
+      queryConfigs[veloxConfigName] =
+          configMapping.toVeloxPropertyValueConverter(propertyOpt.value());
+    }
+  }
+}
+} // namespace
+
+velox::core::QueryConfig toVeloxConfigs(
+    const protocol::SessionRepresentation& session) {
+  // Use base velox query config as the starting point and add Presto session
+  // properties on top of it.
+  auto configs = BaseVeloxQueryConfig::instance()->values();
+
+  // Firstly apply Presto system properties to Velox query config.
+  updateFromSystemConfigs(configs);
+
+  // Secondly apply and possibly override with Presto session properties.
+  updateFromSessionConfigs(session, configs);
+
+  // Finally apply special case configs.
+  updateVeloxConfigsWithSpecialCases(configs);
```

Comment:
We should merge L176 and L182?

Reply:
Thanks @kewang1024 , line 182 is temporary special overrides that should be called after all config and session are applied. But that's okay. I'm planning to have followup to clean the system config as well. The logic here will become more clear.